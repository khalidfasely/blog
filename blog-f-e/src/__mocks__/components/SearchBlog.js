//function SearchBlog({ children }){
//    return <>{children}</>;
//  }
//  
//  export default SearchBlog;

export default jest.mock('../../components/SearchBlog', () => () => 'SearchBlog');
