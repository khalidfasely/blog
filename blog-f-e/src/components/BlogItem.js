import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import ReactMarkdown from "react-markdown";
import { startDeleteBlog } from '../actions/blogs';

const BlogItem = ({ blog, uname, startDeleteBlog }) => {
  const [rModalOpen, setRModalOpen] = useState(false);
  const [eModalOpen, setEModalOpen] = useState(false);

  const deleteBlog = () => {
    startDeleteBlog(blog.id);
    setRModalOpen(false);
  }

  return (
    <div>
      <Link to={`/blog/${blog.id}`}>{blog.title}</Link> by <Link to={`/user/${blog.created_by.id}`}>{blog.created_by.username}</Link> -- {blog.created_at}
      <div>{blog.description}</div>
      <div>{blog.likes} -- {blog.dislikes}</div>
      {
        uname===blog.created_by.username &&
        <div>
          <button onClick={() => setRModalOpen(true)}>X</button>
          <button onClick={() => setEModalOpen(true)}>Edit</button>
        </div>
      }
      <Modal
        isOpen={rModalOpen}
        contentLabel="Delete Blog"
        closeTimeoutMS={100}
      >
        <h5>If you delete this blog there is no way to return it.</h5>
        <button onClick={deleteBlog}>Delete Blog</button>
        <button onClick={() => setRModalOpen(false)}>X</button>
      </Modal>
    </div>
  )
};//closeTimeoutMS={200}

const mapStateToProps = (state) => ({
  uname: state.auth.uname
});

const mapDispatchToProps = (dispatch) => ({
  startDeleteBlog: (bid) => dispatch(startDeleteBlog(bid))
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogItem);

//<ReactMarkdown>{blog.content}</ReactMarkdown>