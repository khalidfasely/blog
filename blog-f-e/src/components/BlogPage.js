import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { startAddComment, startLikeBlog, startSetBlog, startUnlikeBlog } from '../actions/blogPage';
import Comment from './Comment';
import { Link } from 'react-router-dom';
import { l_blog, s_blog, u_blog, u_s_blog } from '../actions/auth';
import { startSaveBlog, startUnsaveBlog } from '../actions/savedBlogs';
import ModalEdit from './ModalEdit';
import previewContent from '../functions/previewContent';

const BlogPage = (props) => {
    const [renderBlog, setRenderBlog] = useState(false);
    const [blog, setBlog] = useState(undefined);
    const [newComment, setNewComment] = useState('');
    const [ buttonDis, setButtonDis ] = useState(false);
    const [ sButtonDis, setSButtonDis ] = useState(false);
    const [eModalOpen, setEModalOpen] = useState(false);
    const [ previewBlog, setPreviewBlog ] = useState(undefined);
    const disableButton = () => {
        setButtonDis(true);
        setTimeout(() => {
            setButtonDis(false);
        }, 200);
    };
    const disableSButton = () => {
        setSButtonDis(true);
        setTimeout(() => {
            setSButtonDis(false);
        }, 200);
    };
    const blogDidAlreadyLoad = () => {
        let blogInList = undefined;
        props.blogList.map((blogItem) => {
            if(blogItem.id === parseInt(props.match.params.bid)){
                blogInList = blogItem;
            };
        });
        return blogInList;
    }
    useEffect(() => {
        const blog = blogDidAlreadyLoad();
        if (blog) {
            setBlog(blog);
            document.title = `${blog.title} (@${blog.created_by.username}) - Blog`;
            setRenderBlog(true);
            setPreviewBlog(previewContent(blog.content.substring(0, 300)));
            // Render the blog content.. from Redux
            // Otherwise Fecth it from the backend
            // Store it on Redux
            // Then render the content from Redux
        } else {
            props.startSetBlog(parseInt(props.match.params.bid)).then((result) => {
                setBlog({
                    ...result.blog,
                    comments: result.comments
                });
                document.title = `${result.blog.title} (@${result.blog.created_by.username}) - Blog`;
                setRenderBlog(true);
                setPreviewBlog(previewContent(result.blog.content.substring(0, 300)));
                //this.setState({ renderBlog: true });
            });
        }
        //return () => {
        //    setPreviewBlog(undefined);
        //}
    }, [])
    const onCommentChange = (e) => {
        const newComment = e.target.value;
        setNewComment(newComment);
    }

    const onFormSubmit = (e) => {
        e.preventDefault();
        if (newComment.replace(/\s/g, '')) {    
            props.startAddComment(parseInt(props.match.params.bid), newComment)//.then((result) => {

//            })
            setNewComment('');
        }
    }

    const like_b = () => {
        disableButton();
        props.startLikeBlog(parseInt(props.match.params.bid)).then(() => {
            props.l_blog(parseInt(props.match.params.bid));
            setBlog({
                ...blog,
                likes: blog.likes + 1
            });
        });
    };

    const unlike_b = () => {
        disableButton();
        props.startUnlikeBlog(parseInt(props.match.params.bid)).then(() => {
            props.u_blog(parseInt(props.match.params.bid));
            setBlog({
                ...blog,
                likes: blog.likes - 1
            });
        });
    };

    const save_b = () => {
        disableSButton();
        props.startSaveBlog(parseInt(props.match.params.bid)).then(() => {
            props.s_blog(parseInt(props.match.params.bid));
        });
    };

    const unsave_b = () => {
        disableSButton();
        props.startUnsaveBlog(parseInt(props.match.params.bid)).then(() => {
            props.u_s_blog(parseInt(props.match.params.bid));
        });
    };

    if(renderBlog) {
        if (blog.isDefault) {
            return <div>This Blog Not Available Now. <Link to='/'>Back Home</Link></div>
        } else if (props.isPreview) {
            return (
                <div>
                    <h1>{blog.title}</h1>
                    <p>{blog.created_at}</p>
                    <p>{blog.created_by.username}</p>
                    <p>{blog.category}</p>
                    <p>{blog.description}</p>
                    <ReactMarkdown>{`${previewBlog}...`}</ReactMarkdown>
                    <Link to={`/blog/${blog.id}`}>See More</Link>
                    <p>{blog.likes}</p>
                    <p>{blog.dislikes}</p>
                    <div>
                        {
                            props.uname && (
                                props.blogsLiked.includes(props.match.params.bid) ?
                                <button disabled={buttonDis} onClick={unlike_b}>Unlike</button> :
                                <button disabled={buttonDis} onClick={like_b}>Like</button>
                                )
                        }
                    </div>
                    <div>
                        {
                            props.uname && (
                                props.blogsSaved.includes(props.match.params.bid) ?
                                <button disabled={sButtonDis} onClick={unsave_b}>Unsave</button> :
                                <button disabled={sButtonDis} onClick={save_b}>Save</button>
                            )
                        }
                    </div>
                    <div>{blog.comments.length} Comment(s)</div>
                </div>
            )
        }
        return (
            <div>
                <h1>{blog.title}</h1>
                <p>{blog.created_at}</p>
                <p>{blog.created_by.username}</p>
                <p>{blog.category}</p>
                <p>{blog.description}</p>
                <ReactMarkdown>{blog.content}</ReactMarkdown>
                <p>{blog.likes}</p>
                <p>{blog.dislikes}</p>
                {
                    props.uname && (
                        props.uname === blog.created_by.username &&
                        <button onClick={() => setEModalOpen(true)}>Edit</button>
                    )
                }
                <ModalEdit eModalOpen={eModalOpen} blog={blog} setEModalOpen={setEModalOpen} />
                <div>
                    {
                        props.uname && (
                            props.blogsLiked.includes(props.match.params.bid) ?
                            <button disabled={buttonDis} onClick={unlike_b}>Unlike</button> :
                            <button disabled={buttonDis} onClick={like_b}>Like</button>
                        )
                    }
                </div>
                <div>
                    {
                        props.uname && (
                            props.blogsSaved.includes(props.match.params.bid) ?
                            <button disabled={sButtonDis} onClick={unsave_b}>Unsave</button> :
                            <button disabled={sButtonDis} onClick={save_b}>Save</button>
                        )
                    }
                </div>
                <div>
                    <h5>Comments:</h5>
                    {
                        blog.comments.length !== 0 ?
                        blog.comments.map(comment => <Comment key={comment.id} blogId={blog.id} comment={comment} />) :
                        <div>-NO COMMENTS-</div>
                    }
                </div>
                <div>
                    <h5>Add comments:</h5>
                    {
                        props.uname ?
                        <form onSubmit={onFormSubmit}>
                            <textarea
                            placeholder='Comment'
                            onChange={onCommentChange}
                            value={newComment}
                            />
                            <button disabled={!newComment.replace(/\s/g, '')}>Add Comment</button>
                        </form> :
                        <h5><Link to='/login'>Login</Link> or <Link to='/register'>Sign In</Link> to Add a Comment!</h5>
                    }
                </div>
            </div>
        )
    } else {
        return <div>Blog Page {props.match.params.bid}</div>
    };
};

const mapStateToProps = (state) => ({
    blogList: state.blogPage,
    uname: state.auth.uname,
    blogsLiked: state.auth.blogsLiked,
    blogsSaved: state.auth.blogsSaved
});

const mapDispatchToProps = (dispatch) => ({
    startSetBlog: (id) => dispatch(startSetBlog(id)),
    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment })),
    startLikeBlog: (id) => dispatch(startLikeBlog(id)),
    startUnlikeBlog: (id) => dispatch(startUnlikeBlog(id)),
    l_blog: (id) => dispatch(l_blog(id)),
    u_blog: (id) => dispatch(u_blog(id)),
    s_blog: (id) => dispatch(s_blog(id)),
    u_s_blog: (id) => dispatch(u_s_blog(id)),
    startSaveBlog: (id) => dispatch(startSaveBlog(id)),
    startUnsaveBlog: (id) => dispatch(startUnsaveBlog(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);

//import React, { useEffect, useState } from 'react';
//import { connect } from 'react-redux';
//import ReactMarkdown from 'react-markdown';
//import { startAddComment, startSetBlog } from '../actions/blogPage';
//import Comment from './Comment';
//
//const BlogPage = (props) => {
//    const [renderBlog, setRenderBlog] = useState(false);
//    const [blog, setBlog] = useState(undefined);
//    const [newComment, setNewComment] = useState('');
//
//    const blogDidAlreadyLoad = () => {
//        let blogInList = undefined;
//        props.blogList.map((blogItem) => {
//            if(blogItem.id === parseInt(props.match.params.bid)){
//                blogInList = blogItem;
//            };
//        });
//        return blogInList;
//    }
//    useEffect(() => {
//        const blog = blogDidAlreadyLoad();
//        if (blog) {
//            setBlog(blog);
//            setRenderBlog(true);
//            // Render the blog content.. from Redux
//            // Otherwise Fecth it from the backend
//            // Store it on Redux
//            // Then render the content from Redux
//        } else {
//            props.startSetBlog(parseInt(props.match.params.bid)).then((result) => {
//                setBlog({
//                    ...result.blog,
//                    comments: result.comments
//                });
//                setRenderBlog(true);
//                //this.setState({ renderBlog: true });
//            });
//        }
//    }, [])
//
//    const onCommentChange = (e) => {
//        const newComment = e.target.value;
//        setNewComment(newComment);
//    }
//
//    const onFormSubmit = (e) => {
//        e.preventDefault();
//        const id = parseInt(props.match.params.bid);
//        if (newComment.replace(/\s/g, '')) {    
//            props.startAddComment(id, newComment)//.then((result) => {
//
////            })
//            setNewComment('');
//        }
//    }
//
//    if(renderBlog) {
//        return (
//            <div>
//                <h1>{blog.title}</h1>
//                <p>{blog.created_at}</p>
//                <p>{blog.created_by.username}</p>
//                <p>{blog.category}</p>
//                <p>{blog.description}</p>
//                <ReactMarkdown>{blog.content}</ReactMarkdown>
//                <p>{blog.likes}</p>
//                <p>{blog.dislikes}</p>
//                <div>
//                    <h5>Comments:</h5>
//                    {
//                        blog.comments.length !== 0 ?
//                        blog.comments.map(comment => <Comment key={comment.id} comment={comment} />) :
//                        <div>-NO COMMENTS-</div>
//                    }
//                </div>
//                <div>
//                    <h5>Add comments:</h5>
//                    <form onSubmit={onFormSubmit}>
//                        <textarea
//                          placeholder='Comment'
//                          onChange={onCommentChange}
//                          value={newComment}
//                        />
//                        <button disabled={!newComment.replace(/\s/g, '')}>Add Comment</button>
//                    </form>
//                </div>
//            </div>
//        )
//    } else {
//        return <div>Blog Page {props.match.params.bid}</div>
//    };
//};
//
//const mapStateToProps = (state) => ({
//    blogList: state.blogPage
//});
//
//const mapDispatchToProps = (dispatch) => ({
//    startSetBlog: (id) => dispatch(startSetBlog(id)),
//    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment }))
//});
//
//export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);