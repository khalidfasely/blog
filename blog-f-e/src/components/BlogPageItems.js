import React from 'react';
import { Link } from 'react-router-dom';
import MarkdownBlog from './MarkdownBlog';

export const BlogPageItems = ({ blog, isPreview }) => {
    return (
        <div className='blog-items__container'>
            {
                blog && (
                    <div className='blog-items'>
                        <div className='solid-border-bottom'>
                            <p className='blog-items__category' data-testid='category_item' >{blog.category}</p>
                            <h3 className='blog-items__title' data-testid='title_item' >{blog.title}</h3>
                            <div className='blog-items__user'>
                                <Link to={`/user/${blog.created_by.id}`}>
                                    <p className='blog-items__username' data-testid='user_item' >
                                        {blog.created_by.username}
                                    </p>
                                </Link>
                                <p className='blog-items__time' data-testid='time_item' >{blog.created_at}</p>
                            </div>
                            <p className='blog-items__description' data-testid='description_item' >{blog.description}</p>
                        </div>
                        <MarkdownBlog blogContent={blog.content} />
                        {
                            isPreview &&
                            <div className='blog-items__link-page'>
                                <Link data-testid='blog_link' to={`/blog/${blog.id}`}>
                                    Read More
                                </Link>
                            </div>
                        }
                        <p className='blog-items__likes' data-testid='likes_item' ><span>{blog.likes}</span> People(s) like this Blog</p>
                    </div>
                )
            }
        </div>
    )
};//<p data-testid='dislikes_item' >{blog.dislikes}</p>

export default BlogPageItems;