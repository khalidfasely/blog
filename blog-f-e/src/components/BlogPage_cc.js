import React from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { startAddComment, startLikeBlog, startSetBlog, startUnlikeBlog } from '../actions/blogPage';
import Comment from './Comment';
import { Link } from 'react-router-dom';
import { l_blog, u_blog } from '../actions/auth';

class BlogPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            renderBlog: false,
            blog: undefined,
            newComment: ''
        }
    }
    //state = {
    //    renderBlog: false,
    //    blog: undefined,
    //    newComment: ''
    //}
    blogDidAlreadyLoad = () => {
        let blogInList = undefined;
        this.props.blogList.map((blogItem) => {
            if(blogItem.id === parseInt(this.props.match.params.bid)){
                blogInList = blogItem;
            };
        });
        return blogInList;
    }
    //componentDidUpdate(){
    //    console.log(this.state.blog)
    //}
    componentWillMount(){
        //console.log(this.props.blogsLiked, this.props.match.params.bid, this.props.blogsLiked.includes(this.props.match.params.bid))
        const blog = this.blogDidAlreadyLoad();
        if (blog) {
            this.setState({ blog });
            this.setState({ renderBlog: true });
            // Render the blog content.. from Redux
            // Otherwise Fecth it from the backend
            // Store it on Redux
            // Then render the content from Redux
        } else {
            this.props.startSetBlog(parseInt(this.props.match.params.bid)).then((result) => {
                this.setState({ blog: {
                    ...result.blog,
                    comments: result.comments
                }, renderBlog: true });
                //this.setState({ renderBlog: true });
            });
        }
    }
    componentDidUpdate(props, prevState){
        if (prevState.blog) {
            //if (prevState.blog !== this.state.blog) {
                console.log(prevState.blog, this.state.blog, prevState.blog === this.state.blog);
            //}
        }
        
        //if (prevState.blog !== undefined) {
        //    if (prevState.blog !== this.state.blog) {
        //        console.log("update");
        //    }
        //}
    }

    onCommentChange = (e) => {
        const newComment = e.target.value;
        this.setState({ newComment });
    }

    onFormSubmit = (e) => {
        e.preventDefault();
        if (this.state.newComment.replace(/\s/g, '')) {    
            this.props.startAddComment(parseInt(this.props.match.params.bid), this.state.newComment)//.then((result) => {

//            })
            this.setState({ newComment: '' });
        }
    }

    like_b = () => {
        //fetch(`/data/like_blog/${this.props.match.params.bid}`)
        //.then(res => res.json())
        //.then(result => console.log(result))
        //.catch(er => console.log(er));
        this.props.startLikeBlog(parseInt(this.props.match.params.bid)).then(() => {
            this.props.l_blog(parseInt(this.props.match.params.bid));
        });
    };

    unlike_b = () => {
        this.props.startUnlikeBlog(parseInt(this.props.match.params.bid)).then(() => {
            this.props.u_blog(parseInt(this.props.match.params.bid));
        });
    };

    render(){
        if(this.state.renderBlog) {
            return (
                <div>
                    <h1>{this.state.blog.title}</h1>
                    <p>{this.state.blog.created_at}</p>
                    <p>{this.state.blog.created_by.username}</p>
                    <p>{this.state.blog.category}</p>
                    <p>{this.state.blog.description}</p>
                    <ReactMarkdown>{this.state.blog.content}</ReactMarkdown>
                    <p>{this.state.blog.likes}</p>
                    <p>{this.state.blog.dislikes}</p>
                    <div>
                        {
                            this.props.blogsLiked.includes(this.props.match.params.bid) ?
                            <button onClick={this.unlike_b}>Unlike</button> :
                            <button onClick={this.like_b}>Like</button>
                            //<button onClick={this.unlike_b}>Unlike</button>
                        }
                    </div>
                    <div>
                        <h5>Comments:</h5>
                        {
                            this.state.blog.comments.length !== 0 ?
                            this.state.blog.comments.map(comment => <Comment key={comment.id} blogId={this.state.blog.id} comment={comment} />) :
                            <div>-NO COMMENTS-</div>
                        }
                    </div>
                    <div>
                        <h5>Add comments:</h5>
                        {
                            this.props.uname ?
                            <form onSubmit={this.onFormSubmit}>
                                <textarea
                                placeholder='Comment'
                                onChange={this.onCommentChange}
                                value={this.state.newComment}
                                />
                                <button disabled={!this.state.newComment.replace(/\s/g, '')}>Add Comment</button>
                            </form> :
                            <h5><Link to='/login'>Login</Link> or <Link to='/register'>Sign In</Link> to Add a Comment!</h5>
                        }
                    </div>
                </div>
            )
        } else {
            return <div>Blog Page {this.props.match.params.bid}</div>
        };
    };
};

const mapStateToProps = (state) => ({
    blogList: state.blogPage,
    uname: state.auth.uname,
    blogsLiked: state.auth.blogsLiked
});

const mapDispatchToProps = (dispatch) => ({
    startSetBlog: (id) => dispatch(startSetBlog(id)),
    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment })),
    startLikeBlog: (id) => dispatch(startLikeBlog(id)),
    startUnlikeBlog: (id) => dispatch(startUnlikeBlog(id)),
    l_blog: (id) => dispatch(l_blog(id)),
    u_blog: (id) => dispatch(u_blog(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);