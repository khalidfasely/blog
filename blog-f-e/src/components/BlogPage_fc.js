import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { startAddComment, startSetBlog } from '../actions/blogPage';
import Comment from './Comment';

const BlogPage = (props) => {
    const [renderBlog, setRenderBlog] = useState(false);
    const [blog, setBlog] = useState(undefined);
    const [newComment, setNewComment] = useState('');

    const blogDidAlreadyLoad = () => {
        let blogInList = undefined;
        props.blogList.map((blogItem) => {
            if(blogItem.id === parseInt(props.match.params.bid)){
                blogInList = blogItem;
            };
        });
        return blogInList;
    }
    useEffect(() => {
        const blog = blogDidAlreadyLoad();
        if (blog) {
            setBlog(blog);
            setRenderBlog(true);
            // Render the blog content.. from Redux
            // Otherwise Fecth it from the backend
            // Store it on Redux
            // Then render the content from Redux
        } else {
            props.startSetBlog(parseInt(props.match.params.bid)).then((result) => {
                setBlog({
                    ...result.blog,
                    comments: result.comments
                });
                setRenderBlog(true);
                //this.setState({ renderBlog: true });
            });
        }
    }, [])

    const onCommentChange = (e) => {
        const newComment = e.target.value;
        setNewComment(newComment);
    }

    const onFormSubmit = (e) => {
        e.preventDefault();
        const id = parseInt(props.match.params.bid);
        if (newComment.replace(/\s/g, '')) {    
            props.startAddComment(id, newComment)//.then((result) => {

//            })
            setNewComment('');
        }
    }

    if(renderBlog) {
        return (
            <div>
                <h1>{blog.title}</h1>
                <p>{blog.created_at}</p>
                <p>{blog.created_by.username}</p>
                <p>{blog.category}</p>
                <p>{blog.description}</p>
                <ReactMarkdown>{blog.content}</ReactMarkdown>
                <p>{blog.likes}</p>
                <p>{blog.dislikes}</p>
                <div>
                    <h5>Comments:</h5>
                    {
                        blog.comments.length !== 0 ?
                        blog.comments.map(comment => <Comment key={comment.id} comment={comment} />) :
                        <div>-NO COMMENTS-</div>
                    }
                </div>
                <div>
                    <h5>Add comments:</h5>
                    <form onSubmit={onFormSubmit}>
                        <textarea
                          placeholder='Comment'
                          onChange={onCommentChange}
                          value={newComment}
                        />
                        <button disabled={!newComment.replace(/\s/g, '')}>Add Comment</button>
                    </form>
                </div>
            </div>
        )
    } else {
        return <div>Blog Page {props.match.params.bid}</div>
    };
};

const mapStateToProps = (state) => ({
    blogList: state.blogPage
});

const mapDispatchToProps = (dispatch) => ({
    startSetBlog: (id) => dispatch(startSetBlog(id)),
    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment }))
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);