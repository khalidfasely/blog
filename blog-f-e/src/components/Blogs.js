import React from "react";
import { Link } from "react-router-dom";
import { connect } from 'react-redux';
import BlogsList from "./BlogsList";


const Blogs = ({ blogs }) => {
  return (
    <div>
      <h1>Blogs Section</h1>
      <BlogsList blogs={blogs} />
      <Link to='/profile'>Profile</Link>
    </div>
  )
};

const mapStateToProps = (state) => ({
  blogs: state.blogs.blogs,
});

export default connect(mapStateToProps)(Blogs);



//blogs.map(blog => (
//  <div>
//    <h1>{blog.id}</h1>
//    <div>{blog.title}</div> by {blog.created_by} -- {blog.created_at}
//    <div>{blog.description}</div>
//    <div>{blog.content}</div>
//    <div>{blog.likes}</div>
//    <div>{blog.dislikes}</div>
//  </div>
//))
//}