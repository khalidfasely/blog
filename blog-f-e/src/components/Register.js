import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startRegister } from '../actions/auth';
import { history } from '../router/AppRouter';

const Register = ({ startRegister }) => {
    const [username, setUsername] = useState('');
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [confirmation, setConfirmation] = useState('');
    const [error, setError] = useState('');

    const onUsernameChange = (e) => {
        const username = e.target.value;
        setUsername(username);
    }

    const onEmailChange = (e) => {
        const email = e.target.value;
        setEmail(email);
    }

    const onPasswordChange = (e) => {
        const password = e.target.value;
        setPassword(password);
    }

    const onConfirmationChange = (e) => {
        const confirmation = e.target.value;
        setConfirmation(confirmation);
    }

    const onFormSubmit = (e) => {
        e.preventDefault();

        const availableData = !!(username && email && password && confirmation && password === confirmation);
        
        if (availableData) {
            //console.log('available data!');
            startRegister({ username, email, password, confirmation }).then((result) => {
                if (result.message !== "Register") {
                    setError(result.message);
                } else {
                    setError('');
                    history.push('/');
                }
            });
        } else {
            setError('Make Sure You Fill All Fields AND The Passwords Match!');
        }
    }

    return (
        <div>
            {error && <div>{error}</div>}
            <form onSubmit={onFormSubmit}>
                <input
                  name='username'
                  type='text'
                  autoFocus
                  placeholder='Username'
                  value={username}
                  onChange={onUsernameChange}
                />
                <input
                  name='email'
                  type='email'
                  placeholder='Email'
                  value={email}
                  onChange={onEmailChange}
                />
                <input
                  name='password'
                  type='password'
                  placeholder='Password'
                  value={password}
                  onChange={onPasswordChange}
                />
                <input
                  name='confirmation'
                  type='password'
                  placeholder='Password (Again)'
                  value={confirmation}
                  onChange={onConfirmationChange}
                />
                <button>Submit</button>
            </form>
            Already have an account? <Link to='/login'>Log In here.</Link>
        </div>
    );
};

const mapDispacthToProps = (dispatch) => ({
    startRegister: ({ username, email, password, confirmation }) => dispatch(startRegister({ username, email, password, confirmation }))
});

export default connect(undefined, mapDispacthToProps)(Register);