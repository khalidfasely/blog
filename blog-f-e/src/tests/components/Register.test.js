import React from 'react';
import { Register } from '../../components/Register';
import { render, fireEvent } from '@testing-library/react';

let getByTestId;

beforeEach(() => {
   const component = render(<Register />);
   getByTestId = component.getByTestId;
});

test('Should render Register component correctly', () => {
    const { asFragment } = render(<Register />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render button with correct text', () => {
    const buttonEl = getByTestId('button');
    expect(buttonEl.textContent).toBe('Register');
});

test('Should change username input value', () => {
    const usernameInputEl = getByTestId('username');

    expect(usernameInputEl.value).toBe('');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    expect(usernameInputEl.value).toBe('Username');
});

test('Should change email input value', () => {
    const emailInputEl = getByTestId('email');

    expect(emailInputEl.value).toBe('');

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    expect(emailInputEl.value).toBe('Email@test');
});

test('Should change password input value', () => {
    const passwordInputEl = getByTestId('password');

    expect(passwordInputEl.value).toBe('');

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    expect(passwordInputEl.value).toBe('Password');
});

test('Should change confirmation input value', () => {
    const confirmationInputEl = getByTestId('confirmation');

    expect(confirmationInputEl.value).toBe('');

    fireEvent.change(confirmationInputEl, {
        target: {
            value: 'Confirmation'
        }
    });

    expect(confirmationInputEl.value).toBe('Confirmation');
});

test('Should not submit the form with empty inputs', () => {
    const formEl = getByTestId('form');

    fireEvent.submit(formEl);

    const errorEl = getByTestId('error_message');

    expect(errorEl.textContent).toBe('Make Sure You Fill All Fields AND The Passwords Match!');
});

test('Should not submit the form with one empty input', () => {
    const formEl = getByTestId('form');

    const usernameInputEl = getByTestId('username');
    const emailInputEl = getByTestId('email');
    const passwordInputEl = getByTestId('password');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    fireEvent.submit(formEl);

    const errorEl = getByTestId('error_message');

    expect(errorEl.textContent).toBe('Make Sure You Fill All Fields AND The Passwords Match!');
});

test('Should not submit the form if password and confirmation does not matchs', () => {
    const formEl = getByTestId('form');

    const usernameInputEl = getByTestId('username');
    const emailInputEl = getByTestId('email');
    const passwordInputEl = getByTestId('password');
    const confirmationInputEl = getByTestId('confirmation');

    fireEvent.change(usernameInputEl, {
        target: {
            value: 'Username'
        }
    });

    fireEvent.change(emailInputEl, {
        target: {
            value: 'Email@test'
        }
    });

    fireEvent.change(passwordInputEl, {
        target: {
            value: 'Password'
        }
    });

    fireEvent.change(confirmationInputEl, {
        target: {
            value: 'AnotherPassword'
        }
    });

    fireEvent.submit(formEl);

    const errorEl = getByTestId('error_message');

    expect(errorEl.textContent).toBe('Make Sure You Fill All Fields AND The Passwords Match!');
});
