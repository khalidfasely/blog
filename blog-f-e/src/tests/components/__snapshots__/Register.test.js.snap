// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Register component correctly 1`] = `
<DocumentFragment>
  <div
    class="content-container"
  >
    <form
      class="register-form"
      data-testid="form"
    >
      <div
        class="register-form__lab-inp"
      >
        <div>
          <label
            for="username_new"
          >
            Username:
          </label>
        </div>
        <div>
          <input
            data-testid="username"
            id="username_new"
            maxlength="35"
            name="username"
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="register-form__lab-inp"
      >
        <div>
          <label
            for="email_new"
          >
            Email:
          </label>
        </div>
        <div>
          <input
            data-testid="email"
            id="email_new"
            maxlength="70"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="register-form__lab-inp"
      >
        <div>
          <label
            for="password_new"
          >
            Password:
          </label>
        </div>
        <div>
          <input
            data-testid="password"
            id="password_new"
            maxlength="120"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="register-form__lab-inp"
      >
        <div>
          <label
            for="confirmation_new"
          >
            Confirmation:
          </label>
        </div>
        <div>
          <input
            data-testid="confirmation"
            id="confirmation_new"
            maxlength="120"
            name="confirmation"
            placeholder="Password (Again)"
            type="password"
            value=""
          />
        </div>
      </div>
      <div>
        <button
          data-testid="button"
        >
          Register
        </button>
      </div>
      <div
        class="register-form__link"
      >
        Already have an account? 
        <a>
          Log In here.
        </a>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
